python/plugins/processing/algs/qgis/TopoColors.py

```diff
@@ -0,0 +1,218 @@
+# -*- coding: utf-8 -*-
+
+"""
+***************************************************************************
+    TopoColors.py
+    --------------
+    Date                 : February 2017
+    Copyright            : (C) 2017 by Nyall Dawson
+    Email                : nyall dot dawson at gmail dot com
+***************************************************************************
+*                                                                         *
+*   This program is free software; you can redistribute it and/or modify  *
+*   it under the terms of the GNU General Public License as published by  *
+*   the Free Software Foundation; either version 2 of the License, or     *
+*   (at your option) any later version.                                   *
+*                                                                         *
+***************************************************************************
+"""
+
+__author__ = 'Nyall Dawson'
+__date__ = 'February 2017'
+__copyright__ = '(C) 2017, Nyall Dawson'
+
+# This will get replaced with a git SHA1 when you do a git archive323
+
+__revision__ = '$Format:%H$'
+
+import os
+import operator
+from collections import defaultdict, deque
+
+from qgis.core import (QgsField,
+                       QgsGeometry,
+                       QgsSpatialIndex,
+                       NULL)
+
+from qgis.PyQt.QtCore import (QVariant)
+
+from processing.core.GeoAlgorithm import GeoAlgorithm
+from processing.core.parameters import ParameterVector
+from processing.core.outputs import OutputVector
+from processing.tools import dataobjects, vector
+
+pluginPath = os.path.split(os.path.split(os.path.dirname(__file__))[0])[0]
+
+
+class TopoColor(GeoAlgorithm):
+    INPUT_LAYER = 'INPUT_LAYER'
+    OUTPUT_LAYER = 'OUTPUT_LAYER'
+
+    def defineCharacteristics(self):
+        self.name, self.i18n_name = self.trAlgorithm('Topological coloring')
+        self.group, self.i18n_group = self.trAlgorithm('Cartographic tools')
+        self.tags = self.tr('topocolor,colors,graph,adjacent,assign')
+
+        self.addParameter(ParameterVector(self.INPUT_LAYER,
+                                          self.tr('Input layer'), [dataobjects.TYPE_VECTOR_POLYGON]))
+
+        self.addOutput(OutputVector(self.OUTPUT_LAYER, self.tr('Colored'), datatype=[dataobjects.TYPE_VECTOR_POLYGON]))
+
+    def processAlgorithm(self, feedback):
+        layer = dataobjects.getObjectFromUri(
+            self.getParameterValue(self.INPUT_LAYER))
+
+        fields = layer.fields()
+        fields.append(QgsField('color_id', QVariant.Int))
+
+        writer = self.getOutputFromName(
+            self.OUTPUT_LAYER).getVectorWriter(
+            fields,
+            layer.wkbType(),
+            layer.crs())
+
+        # use a deque so we can drop features as we write them
+        # it's a bit friendlier on memory usage
+        features = deque(f for f in vector.features(layer))
```

Comment:
Actually that code was removed - in the final version I couldn't use the deque.

In general I much prefer to use iterators and tweak the request to optimise these. But it doesn't work for this particular algorithm.

The reason I've gone with this approach (storing all features/attributes) is to avoid multiple fetching of features. For the graph generation we need to loop through all features and compare them against all other features' geometry. This means we have no choice but to store all features in memory for this part of the algorithm.

Then, we will eventually write out all feature attributes for the final output. So it makes sense to fetch them upfront and avoid a second loop over all the features. We fetch features without geometry because these are still written to the output file, just with no color_id set.

In the end my view is that temporarily using more memory is preferable to multiple fetches from the provider.